### ISSUE TITLE
Image sizes scaled down too aggressively

### ISSUE THREAD

**Issue by weiss** [ChatSecure/ChatSecure-iOS#800]
HTTP Upload generally works fine for me using the current beta (yay!).

The only issue I see is that ChatSecure seems to scale down (at least) JPEG images very aggressively (I typically end up with image sizes of 5-15k), so the resulting quality is quite bad.

**Comment by chrisballinger** [comment]
Ah yes, good call. This is a holdover from the OTRDATA days where transfers took a long time and would often fail to complete if they were too large. They should still probably be downscaled somewhat from their original size to reduce mobile data usage, at least by default.

**Comment by LukeLR** [comment]
I'd love if there was a setting for that, so the user can choose the
compression level of his media files. I usually go for uncompressed
media if possible, so I'd love if there could be a setting for that,
too. Telegram has a really nice ui for selecting the compression level
of videos in their iOS app
(https://telegram.org/blog/calls#one-more-thing) maybe something like
this could be implemented in the far future as well?

**Comment by futurealecks** [comment]
Just wanted to add something to this. I took an image with my iPhone 7+ with the latest TestFlight version of CS. The jpg that was sent was 12MB. I sent that image to the Android's Conversations.im and resent the same image back to CS and it was compressed to 505KB. 

Maybe 505KB is aggressive but 12MB is a big chunk of my 4GB data plan :).

**Comment by weiss** [comment]
Yes, I guess it makes sense to let the compression ratio depend on the original size.

**Comment by chrisballinger** [comment]
@futurealecks Hmmm good point. The max is based on what your server reports, so if your server will allow 100 MB attachments, it will probably compress with 85% quality at full resolution.

**Comment by hemee** [comment]
@chrisballinger I think the best idea is to make a setting that allows changing of compression ratio by user (and make some reminder that a server may abort sending if image size exceeds its limits).

**Comment by weiss** [comment]
To be honest, the images I receive from ChatSecure users are now quite large, often several megabytes.

I think it would be nice if compression would result in a JPEG size of a few hundred kilobytes for 'typical' photos.  I guess there would then be less demand to make this configurable.

And I don't think it makes much sense to let the compression ratio depend on the max upload size allowed by the server.  Servers will usually accept file sizes of (at least) several megabytes; and if they don't, they're unusable.  I'd just show an appropriate error message if the file size exceeds the server limit.

**Comment by therob84** [comment]
I can also report images receiving from CS beeing as large as 3-4 MB which is not only for me as receiver but also for the sender in bad receptions areas a very high volume of data.
I even had a lot of sent/receiving errors, because it took to long to send/receive.

Therefor I would also vote for a more efficient compression routine (for me it seems there are even none compression working, as 3-4MB seems to be the original size). Could it be adopted from Conversations, as I am very satisfied by image sizes/quality (150-500kb), in UI you can choose between "never/automatic/always"

@chrisballinger: Could you include any improvement in next version? Do you needs any more helpful comments about this topic?
Thanks and Cheers, Robert

**Comment by chrisballinger** [comment]
There's not a way for us to choose a specific file size, just a quality
value between 0.0-1.0 and a resolution. Originally we over-compressed the
images to some complaints, so then we tried to compress them at medium-high
quality/resolution without going over the server file size limit. It seems
impossible to satisfy everyone without some sort of configuration options,
but I don't really have the time or resources to implement that.

**Comment by therob84** [comment]
@chrisballinger OK, I do understand this, thanks for the reply. Could you maybe provide the quality numbers (0.0-0.1) you used in the versions with the initial post (5-15KB - @weiss: which version?) and the now observed (3-4MB with v4.1)? Going for a number just slightly worse than it is now would be worth a try?

@inputmice: Does conversations maybe uses also the same jpg-compression algorithm and/or you could share the quality number conversations uses (as I find it does a good job)?

**Comment by weiss** [comment]
> It seems impossible to satisfy everyone without some sort of configuration options

FWIW, my assumption would be that you might not satisfy *everyone* by hardcoding whatever Conversations does when set to "automatic", but I think it would work well enough for most users. The commercial messengers I've seen typically end up with similar image sizes (a few hundred kB) without offering a configuration option.

**Comment by weiss** [comment]
> There's not a way for us to choose a specific file size, just a quality value between 0.0-1.0 and a resolution.

Actually, if I read the code correctly, you *do* already have [magic to target a desired file size][1].  The problem seems to be that you're happy as soon as the image size is below the upload file size limit reported by the server.  As far as I can see, if the limit is larger than the original file size, you won't scale the image at all?

Maybe just setting [`maxSize`][2] to `min(dataSize.numBytes, 524288)` (or so) for this loop would already do the trick?

[1]: https://github.com/ChatSecure/ChatSecure-iOS/blob/v4.1.0/ChatSecure/Classes/Controllers/FileTransferManager.swift#L53
[2]: https://github.com/ChatSecure/ChatSecure-iOS/blob/v4.1.0/ChatSecure/Classes/Controllers/FileTransferManager.swift#L52

**Comment by chrisballinger** [comment]
Yeah except you also will need to tune the initial quality factors so it
doesn't loop a bunch of times trying to reach that size. They will need to
be different if you have 8MP camera vs 12MP, but whatever. Something in the
middle should work well enough for most people i guess

**Comment by chrisballinger** [comment]
Try the new build

**Comment by therob84** [comment]
@chrisballinger you are really great in user-feedback-reaction-and-implementation!
I can't test it as I always sit with Conversations at the other side of some friends CS and guessing from some of their complaints and errors what to improve in CS ;-) So somebody else has to test it...I have to wait until the update to the next version this is included....But thanks!!!

**Comment by therob84** [comment]
@chrisballinger: will your above mentioned changes be included into the next release (4.3)?
